src/commands/playground/util.rs

```diff
@@ -198,26 +197,44 @@ pub fn hoise_crate_attributes(code: &str, after_crate_attrs: &str, after_code: &
 /// Utility used by the commands to wrap the given code in a `fn main` if not already wrapped.
 /// To check, whether a wrap was done, check if the return type is Cow::Borrowed vs Cow::Owned
 /// If a wrap was done, also hoists crate attributes to the top so they keep working
-pub fn maybe_wrap(code: &str, result_handling: ResultHandling) -> Cow<'_, str> {
-	if code.contains("fn main") || code.contains("#![no_main]") {
-		return Cow::Borrowed(code);
-	}
+pub fn maybe_wrap(code: &str, result_handling: ResultHandling) -> String {
+	use quote::quote;
+	use syn::{parse::Parse, *};
 
-	// fn main boilerplate
-	let after_crate_attrs = match result_handling {
-		ResultHandling::None => "fn main() {\n",
-		ResultHandling::Discard => "fn main() { let _ = {\n",
-		ResultHandling::Print => "fn main() { println!(\"{:?}\", {\n",
-	};
+	struct Inline {
+		attrs: Vec<Attribute>,
+		exprs: Vec<Stmt>,
+	}
 
-	// fn main boilerplate counterpart
-	let after_code = match result_handling {
-		ResultHandling::None => "}",
-		ResultHandling::Discard => "}; }",
-		ResultHandling::Print => "}); }",
-	};
+	impl Parse for Inline {
+		fn parse(input: parse::ParseStream) -> Result<Self> {
+			let attrs = Attribute::parse_inner(input)?;
+			let exprs = Block::parse_within(input)?;
+			for stmt in &exprs {
+				if let Stmt::Item(Item::Fn(ItemFn { sig, block, .. })) = stmt {
+					if sig.ident.to_string() == "main" && sig.inputs.len() == 0 {
+						return Ok(Self {
+							attrs,
+							exprs: block.stmts.clone(),
+						});
+					}
+				}
+			}
+			Ok(Self { attrs, exprs })
+		}
+	}
 
-	Cow::Owned(hoise_crate_attributes(code, after_crate_attrs, after_code))
+	let Inline { attrs, exprs } = parse_str::<Inline>(code).unwrap();
+	let attrs = attrs.into_iter().map(|x| quote!(#x));
+	let exprs = exprs.into_iter().map(|x| quote!(#x));
+	match result_handling {
+		ResultHandling::None => quote! { #(#attrs)* fn main() { #(#exprs)* } },
+		ResultHandling::Discard => quote! { #(#attrs)* fn main() { _ =  { #(#exprs)* } } },
+		ResultHandling::Print => {
+			quote! { #(#attrs)* fn main() { println!("{:?}", { #(#exprs)* }) } }
```

Comment:
This is unrelated to your PR, but here are some changes I would personally have liked seeing with the bot:

  - Use `|| -> () { ... }()` to make `if true { return 42; }` still work;
  - In the `println!` case:
      - use `#?`-printing;
      - use `::std::`-qualifier on it to be robust to `#[macro_export] macro_rules! println` in the code.

```suggestion
		ResultHandling::Discard => quote!(
		    #(#attrs)*
		    fn main() {
		        _ = || -> () {
		            #(#exprs)*
		        }();
		    }
		),
		ResultHandling::Print => quote!(
		    #(#attrs)*
		    fn main() {
		         ::std::println!("{:#?}", || -> () { #(#exprs)* }());
		    }
		),
```

Reply:
i disagree about the `#?` printing-- i like it printing in `:?` as is, ::std is fine

Reply:
Yeah `#?` is subjective/debatable for sure 👍

Reply:
actually- the closure here doesnt make too much sense, why do you `-> ()`?

Reply:
okay now you can use ![image](https://github.com/rust-community-discord/ferrisbot-for-discord/assets/70787919/cd309480-57af-4b66-b41f-ff3523820a27) to get `#?`.